### PR TITLE
Dvaepp clean merge resolve conflicts

### LINKED ISSUE #5: Decouple modelMaker.py into model initialization and training engine
Would be cleaner and easier to have separate files for handling model initialization (model selection logic, device managment, model init parameter handling etc.) and model training and evaluation (data handling, logging, wandb integration etc.)

### LINKED ISSUE #7: Decouple and remove duplication from rbm.py for the RBM initialization and sampling procedures

### LINKED ISSUE #11: Refactor network.py to decouple basic networks from encoders/decoders

### LINKED ISSUE #17: Implement wandb package
Instead of taking care of logging ourselves, we decided to outsource this to the wandb package. Which steps are necessary to run this?

### LINKED ISSUE #22: Remove data/ and wandb/ directories and rewrite history
The `data/MNIST` subdirectory slipped through `.gitignore` and is now part of the repo's history. These binary files should be removed. There's an open-source tool available to do that called `bfg` (https://rtyley.github.io/bfg-repo-cleaner/).

At the end of the cleaning process, we need to delete our local clones and clone a fresh, cleaned version from upstream. Let's do that once we have committed all local changes.

### PR DESCRIPTION
**Pull request by ezeeEric**
Hey @abhishekabhishek ,

so this is it, the cleaned up merge of our devs. I will comment inline for explanations and important points. Overall, I took in all your changes and made some modifications where we had duplicated work.

Please let me know if you have comments!

Caveats:

- DVAEPP model failing in MixtureExp distribution
- the RBM refactoring changed a few things and needs further testing - will do that.

Addresses the following issues:

- Resolves #22 
- Closes #17 
- Closes #11 
- Closes #7 
- Closes #5

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,15 +1,27 @@
...
 .vscode/
 output/
 sources/
 .gif
 venv_divae/
+outputs/
+wandb/
+data/MNIST/
+MNIST/
+
```
Comment (ezeeEric) [inline .gitignore:16]: I created this block for our manual additions to the gitignore.

File: configs/data/mnist.yaml [hunk 2]
```diff
@@ -3,9 +3,10 @@ data_type: mnist
 binarise_dataset: threshold
 # load_data_from_pkl: 0
 
-#user dependent
-pre_processed_input_file: /Users/drdre/inputz/MNIST/preprocessed/full.pkl
-mnist_input: /Users/drdre/inputz/
+# Automatic download applicable if MNIST files could not be found
+# Use top-level directory path for the repo as mnist_input
+pre_processed_input_file: 0
+mnist_input: /home/akajal/DiVAE
```
Comment (ezeeEric) [inline configs/data/mnist.yaml:9]: The default config is now pointing to your input path. In one of the coming updates, I will add the possibility of user dependent default config files.

File: configs/model/divaepp.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+model_type: DiVAEPP
+activation_fct: tanh
+beta_smoothing_fct: 4
+decoder_hidden_nodes: 
+  - 200
+  - 200
+n_encoder_layer_nodes: 200
+n_encoder_layers: 2
+n_latent_hierarchy_lvls: 4
+n_latent_nodes: 100
```
Comment (ezeeEric) [inline configs/model/divaepp.yaml:11]: This is the configuration file for your model. Simply specify `divaepp` in the top-level config `configs/config.yaml` and the modelCreator (see below) will load the DiVAEPP model and set the parameters from here.

File: engine/engine.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
+"""
```
Comment (ezeeEric) [inline engine/engine.py:1]: This is part of refactoring the modelmaker. We now have an `engine` module which steers the training/testing/evaluation procedure independent of the model creation.

File: models/autoencoders/autoencoder.py [hunk 5]
```diff
@@ -67,7 +67,8 @@ def forward(self, input_data):
     
     def loss(self, input_data, fwd_out):
         loss=self._loss_fct(fwd_out.output_data, input_data.view(-1,self._flat_input_size), reduction='sum')
-        return loss
+        return {"loss":loss}
```
Comment (ezeeEric) [inline models/autoencoders/autoencoder.py:70]: This is how wandb starts logging metrics, right? Should I add this change to the other models too? Maybe we can automatise this using the inheritance structure we have, i.e. adding that to the `autoencoderbase`.

File: models/autoencoders/dvaepp.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
+"""
```
Comment (ezeeEric) [inline models/autoencoders/dvaepp.py:1]: I changed this file slightly wrt your original version to make it compatible with the refactoring. Let me know if there are questions or issues.

File: models/modelCreator.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
+"""
```
Comment (ezeeEric) [inline models/modelCreator.py:1]: The second part of the modelMaker refactoring - this is the class for our model handling.

File: utils/dists/MixtureExp.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        """
+        - ICDF of mixture of exponential distributions (Eq. 3, DVAE++)
+        Returns:
+            zeta: approximate post samples
+        """
+        q = torch.sigmoid(self.logits)
+        q = torch.clamp(q, min=1e-7, max=1.-1e-7)
+        
+        rho = torch.rand(q.size())
+        b = (rho + torch.exp(-self.beta)*(q - rho))/((1. - q) - 1.)
```
Comment (ezeeEric) [inline utils/dists/MixtureExp.py:29]: This is currently failing, as `self.beta` is an `int` but should be a `torch.int`. Probably just need to cast the type inline.

**Comment by abhishekabhishek** [comment]
@ezeeEric Should be OK for merging into the base branch

**Pull request merged**